### ISSUE TITLE
[clang] AST doesn't track source locations for where the attribute list appears

### ISSUE THREAD

**Issue by v1nh1shungry** [llvm/llvm-project#59935]
When I tried to implement inserting a fix-it into the end of a `VarDecl`, I ran into a problem in the following case:

```cpp
const int var [[clang::annotate("")]];
//                                   ^ insert fix-it here
```

Because I have no way to figure out where the attribute list ends.

Here's the reply from @AaronBallman

> We keep all of the attribute AST nodes with the declaration, and each attribute knows its own source range, but we *don't* keep the source locations for where the full attribute list appears. e.g.,
> ```cpp
> __attribute__((foo)) const int var [[bar, baz]];
> ```
> `var` will have three attributes associated with it, but the only source location information you have access to is for the `foo`, `bar`, and `baz` tokens. Each of those attributes also keeps track of what syntax was used, so you could tell that `foo` was a GNU-style attribute while `bar` and `baz` were C++. But we don't keep enough information about `bar` and `baz` being part of the same attribute list or whether that attribute list is leading or trailing. You have to calculate all of this yourself and some of it might not even be possible to calculate (for example, the user could have done `__attribute__((foo)) const int var [[bar]] [[baz]];` and the AST would come out the same as the previous example).

The original discussion lies in https://reviews.llvm.org/D139705

**Comment by erichkeane** [comment]
Aaron also suggested:

>>something for us to consider is how to improve this in the AST. We run into problems when doing AST pretty printing because of attributes for these same reasons. It seems to me that the Decl/Stmt/TypeLoc nodes need to keep a list of source ranges for attributes around (since there can be multiple ranges for declarations). WDYT?

I suspect that is the right answer.  Decl/Stmt (TypeLoc?) just needs to store a list of SourceRange objects.  From there, we can do comparisons to figure out which attribute is in which (since pretty-printing/fixups are the only time we need this information).  

Unfortunately, this is likely a LOT of additional data on Decl/Stmt/TypeLoc.  I MIGHT instead suggest we add this to the attributes.  That is, each Attr object contains the location of its leading `[[` and trailing `]]`.  While this is MORE information to store on a per-attribute basis, I suspect that we overall will have LESS information, since most Stmts/Decls don't have attributes, so I think it is an overall improvement.  @AaronBallman  WDYT?

**Comment by AaronBallman** [comment]
> Unfortunately, this is likely a LOT of additional data on Decl/Stmt/TypeLoc. I MIGHT instead suggest we add this to the attributes. That is, each Attr object contains the location of its leading `[[` and trailing `]]`. While this is MORE information to store on a per-attribute basis, I suspect that we overall will have LESS information, since most Stmts/Decls don't have attributes, so I think it is an overall improvement. @AaronBallman WDYT?

Hmmm, the problem with storing it on the attribute is 1) it's duplicate information in the case of `[[foo, bar]]` because both attributes will retain the info about the open and close braces, 2) there are a *lot* of attributes (especially in header files) and so the overhead might still be quite high for the benefit gained. However, it could also be a reasonable approach. We also might want to consider a side table of attributes to SourceRanges if we're worried about storing duplicate information.

**Comment by erichkeane** [comment]
Yeah, I considered those duplicates.  I'd imagine it actually wouldn't be much of a problem in header files, which typically do a bunch of macro-magic on a per-attribute-if-supported basis, which means each would have their own leading/trailling bracket pairs.

That said, storing it on the Decl (which means no trailing storage, since it isn't a 'final' class), means storing AT LEAST a `SourceRange` worth of data (which is 2x 32 bit numbers), since you need at least a pointer.  AND most declarations don't have an attribute I'd imagine, so that is a pretty high cost to pay.

**Comment by AaronBallman** [comment]
Yeah, and we've discussed bumping our SourceLocation to be 64-bits, so the expense can go up in the future as well. I'm leaning more towards a side table; then AST nodes don't get larger, but ASTContext does (that's where I'd imagine we'd want this info to live, but maybe Sema is reasonable as well).

**Comment by erichkeane** [comment]
> Yeah, and we've discussed bumping our SourceLocation to be 64-bits, so the expense can go up in the future as well. I'm leaning more towards a side table; then AST nodes don't get larger, but ASTContext does (that's where I'd imagine we'd want this info to live, but maybe Sema is reasonable as well).

A 64 bit `SourceLocation` pair is, IMO, still an order of magnitude less data than putting something on every `Decl`.

Side-tables are yucky for serialization/modules, otherwise I'd be fine with it I think, depending.  We'd at least have to make sure the attributes themselves serialize enough information to rebuild it when reading the module.

**Comment by AaronBallman** [comment]
Good point about modules/PCH making side tables kind of painful. At the end of the day, the solution needs to have overhead that's acceptable enough for the functionality it enables.

**Comment by tomasz-kaminski-sonarsource** [comment]
I believe that I would be impossible for the AST to provide all source locations involved in the C++ construct required for the fix-it and other automatic code edits. Given that, retokenizing of the source is unavoidable at some point for such tools. For example, I think it would be unreasonable to expect AST to provide the location of the `const` or `*`/`&` for all types. 

However, we can still improve the situation. In this concrete case, including attributes in the source range of the declaration may be indeed a bit academic question, but doing so has pragmatic benefits, as:
* the body or initializer in the definition can always be inserted after `getEnd()` (addressing the original issue)
* it would be possible to now remove the declaration by simply removing it `getSourceRange()`
const
* changing type to reference would be equivalent to inserting `&` after `getEnd()` for type

As for the inevitable question of the quality of impact on a diagnostic that wants to refer to declarations, I think this will not have a negative effect. As we already include the whole function body and class definition in the rage, the common practice refers to raising diagnostic on tokens constituting the declared name.

Of course, handling all of the above is always possible by reanalyzing the source after the `getEnd()` token (as currently reported) for the presence of attributes. However, such an approach does not provide a bound for parsing other than the end of the file.

To summarize, giving source range for each attribute may be overkill, and requiring tokenization if needed seems fine, given that we already require them in another context. However, such action should not be needed when the task is the removal of the declaration of add initializer. But changing `getEnd()` to include attributes is sufficient for that.

**Comment by tomasz-kaminski-sonarsource** [comment]
I believe that I would be impossible for the AST to provide all source locations involved in the C++ construct required for the fix-it and other automatic code edits. Given that, retokenizing of the source is unavoidable at some point for such tools. For example, I think it would be unreasonable to expect AST to provide the location of the `const` or `*`/`&` for all types. 

However, we can still improve the situation. In this concrete case, including attributes in the source range of the declaration may be indeed a bit academic question, but doing so has pragmatic benefits, as:
* the body or initializer in the definition can always be inserted after `getEnd()` (addressing the original issue)
* it would be possible to now remove the declaration by simply removing it `getSourceRange()`
const
* changing type to reference would be equivalent to inserting `&` after `getEnd()` for type

As for the inevitable question of the quality of impact on a diagnostic that wants to refer to declarations, I think this will not have a negative effect. As we already include the whole function body and class definition in the rage, the common practice refers to raising diagnostic on tokens constituting the declared name.

Of course, handling all of the above is always possible by reanalyzing the source after the `getEnd()` token (as currently reported) for the presence of attributes. However, such an approach does not provide a bound for parsing other than the end of the file.

To summarize, giving source range for each attribute may be overkill, and requiring tokenization if needed seems fine, given that we already require them in another context. However, such action should not be needed when the task is the removal of the declaration of add initializer. But changing `getEnd()` to include attributes is sufficient for that.
